src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -105,6 +106,17 @@ public partial class Ping
                     {
                         fixed (byte* bytesPtr = &receiveBuffer[0])
                         {
+                            if (isIpv4)
+                            {
+                                // Determine actual size of IP header
+                                byte ihl = (byte)(bytesPtr[0] & 0x0f); // Internet Header Length
+                                ipHeaderLength = 4 * ihl;
+                                if (bytesReceived - ipHeaderLength < IcmpHeaderLengthInBytes)
+                                {
+                                    continue; // Not enough bytes to reconstruct actual IP header + ICMP header.
```

Comment:
The code currently assumes that an entire ICMP packet will be read into the buffer in the `ReceiveFromAsync` call, and if the buffer doesn't end up having what we need, then we retry. Thinking about it more, that is probably not valid, although I'm not sure of what the best practices are to handle it.